### PR TITLE
Revive library with product, coproduct, compose

### PR DESCRIPTION
**Pull request by garyb**
As per the functor-compose discussion. If we do want to do it this way.

/cc @paf31 @natefaubion

### CODE REVIEW

File: LICENSE [hunk 1]
```diff
@@ -0,0 +1,20 @@
+The MIT License (MIT)
+
+Copyright (c) 2014 PureScript
```
Comment (paf31) [inline LICENSE]: Might want to update the year here.

**Comment by paf31** [comment]
Looks great! We could probably add a `Monad` instance for `Compose` too, although actually there are at least two choices there, so maybe not.

**Comment by natefaubion** [comment]
> We could probably add a Monad instance for Compose too, although actually there are at least two choices there, so maybe not.

I don't think this is possible in general, right? That's why we have transformers.

**Comment by garyb** [comment]
> I don't think this is possible in general, right? That's why we have transformers.

That was my thinking too.

**Comment by paf31** [comment]
The composition of a `Monad` with either a `Traversable` `Monad` or a `Distributive` is a `Monad`.

_Edit_: see http://web.cecs.pdx.edu/~mpj/pubs/RR-1004.pdf and https://www.reddit.com/r/haskell/comments/4nc19n/cross_section_of_the_monad_and_traversable_class/d43y7ss for details, but it basically amounts to swapping the `m` and `n` in the middle of `join :: m (n (m (n a))) -> m (n a)`.

**Comment by paf31** [comment]
To be clear, I don't think we should implement one and not the other, and since they have the same instance head, we can't implement both.

**Comment by garyb** [comment]
Yeah, I was going to suggest we leave it out in that case, and perhaps introduce newtypes for both varieties at some point.

**Comment by paf31** [comment]
I think `inject` and `free` are the only reverse dependencies so far. We can now look up reverse dependencies using the package set :smile:

**Comment by garyb** [comment]
Yeah, those are the only ones I know of too :)

**Comment by garyb** [comment]
Do you want to add `Day` / `Pairing` before I make the new release?

**Comment by paf31** [comment]
No that's ok, we can add it later.

**Pull request merged**